src/read/macinfo.rs

```diff
@@ -0,0 +1,290 @@
+    /// # let offset = DebugMacinfoOffset(0);
+    /// # let read_section_somehow = || &buf;
+    /// # let debug_macinfo_offset_somehow = || offset;
+    /// let debug_macinfo = DebugMacInfo::new(read_section_somehow(), LittleEndian);
+    /// for macinfo in debug_macinfo.get_macinfo(debug_macinfo_offset_somehow()) {
+    ///     println!("Found macro info {:?}", macinfo);
+    /// }
+    /// ```
+    pub fn get_macinfo(
+        &self,
+        offset: DebugMacinfoOffset<R::Offset>,
+    ) -> impl Iterator<Item = DebugMacInfoItem<R>> {
+        let mut input = self.section.clone();
+        if let Ok(_) = input.skip(offset.0) {
+            DebugMacInfoIterator {
+                input_reader: input,
+                finished: false,
+            }
+        } else {
+            // If the offset is invalid, the iterator is immediately finished.
+            DebugMacInfoIterator {
+                input_reader: input,
+                finished: true,
+            }
+        }
+    }
+}
+
+impl<T> DebugMacInfo<T> {
+    /// Create a `DebugMacInfo` section that references the data in `self`.
+    ///
+    /// This is useful when `R` implements `Reader` but `T` does not.
+    ///
+    /// Used by `DwarfSections::borrow`.
+    pub fn borrow<'a, F, R>(&'a self, mut borrow: F) -> DebugMacInfo<R>
+    where
+        F: FnMut(&'a T) -> R,
+    {
+        borrow(&self.section).into()
+    }
+}
+
+impl<R> Section<R> for DebugMacInfo<R> {
+    fn id() -> SectionId {
+        SectionId::DebugMacinfo
+    }
+
+    fn reader(&self) -> &R {
+        &self.section
+    }
+}
+
+impl<R> From<R> for DebugMacInfo<R> {
+    fn from(macinfo_section: R) -> Self {
+        DebugMacInfo {
+            section: macinfo_section,
+        }
+    }
+}
+
+/// an Entry in the `.debug_macinfo` section.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DebugMacInfoItem<R> {
+    /// The end of the unit.
+    EndOfUnit,
+    /// A macro definition.
+    Define {
+        /// The line number where the macro is defined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
+        macro_text: R,
+    },
+    /// A macro undefinition.
+    Undef {
+        /// The line number where the macro is undefined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
+        macro_text: R,
+    },
+    /// The start of a file.
+    StartFile {
+        /// line number of the source line on which the inclusion macro directive occurred
+        line_number: u64,
+        /// the index into the line number table of the compilation unit (LEB128).
+        file_number_index: u64,
+    },
+    /// The end of a file.
+    EndFile,
+    /// A vendor-specific extension.
+    VendorExt {
+        /// a numeric constant (LEB128), whose meaning is vendor specific.
+        numeric_constant: u64,
+        /// a null-terminated string, whose meaning is vendor specific.
+        string_constant: R,
```

Comment:
```suggestion
        /// A numeric constant whose meaning is vendor specific.
        numeric: u64,
        /// A string whose meaning is vendor specific.
        string: R,
```
Field names are needlessly verbose. The DWARF encoding (LEB128 and null-terminated) is irrelevant here. Capitalize the start of comments.

Similarly for other variants.